src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -79,7 +89,212 @@ struct C_EncryptedObjectReadRequest : public Context {
+      }
+      if (aligns.second != 0) {
+        extents.emplace_back(object_off + data->length(),
+                             aligns.second);
+      }
+
+      if (cmp_data != nullptr) {
+        extents.emplace_back(object_off, cmp_data->length());
+      }
+
+      read_req = new C_UnalignedObjectReadRequest<I>(
+              image_ctx, crypto, object_no, &extents, io_context,
+              0, 0, parent_trace, &version, 0, this);
+
+      write_req_sender = [image_ctx, object_no, aligned_object_off, io_context,
+                          op_flags, write_flags, assert_version, journal_tid,
+                          &parent_trace,
+                          this] (std::optional<uint64_t> version) mutable {
+        bool exclusive = write_flags & io::OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE;
+        if (!version.has_value()) {
+          write_flags |= io::OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE;
+          version = assert_version;
+        } else if (exclusive) {
+          onfinish->complete(-EEXIST);
+          return;
+        } else if (assert_version.has_value() &&
+                   version.value() != assert_version.value()) {
+          if (assert_version.value() < version.value()) {
+            onfinish->complete(-ERANGE);
+          } else {
+            onfinish->complete(-EOVERFLOW);
+          }
+          return;
+        }
+
+        auto ctx = new LambdaContext(
+                [this, exclusive, assert_version](int r) {
+                  if (r == -ERANGE && !assert_version.has_value()) {
+                    r = -ERESTART;
+                  } else if (r == -EEXIST && !exclusive) {
+                    r = -ERESTART;
+                  }
+                  onfinish->complete(r);
+                });
+
+        auto write_req = io::ObjectDispatchSpec::create_write(
+                image_ctx, io::OBJECT_DISPATCH_LAYER_NONE, object_no,
+                aligned_object_off, std::move(*aligned_data), io_context,
+                op_flags, write_flags, version,
+                journal_tid == nullptr ? 0 : *journal_tid, parent_trace, ctx);
+        write_req->send();
+      };
+    }
+
+    void send() {
+      read_req->send();
+    }
+
+    void finish(int r) override {
+      ldout(cct, 20) << "r=" << r << dendl;
+
+      bool no_obj = false;
+      if (r == -ENOENT) {
+        no_obj = true;
+      } else if (r < 0) {
+        onfinish->complete(r);
+        return;
+      }
+
+      if (cmp_data != nullptr) {
+        auto& read_bl = extents.back().bl;
+        std::optional<uint64_t> found_mismatch = std::nullopt;
+        if (no_obj) {
+          read_bl.append_zero(cmp_data->length());
+        }
+        if (read_bl.length() != cmp_data->length()) {
+          ceph_assert(read_bl.length() < cmp_data->length());
+          found_mismatch = std::make_optional(read_bl.length());
+        } else {
+          for (uint64_t idx = 0; idx < cmp_data->length(); ++idx) {
+            if ((*cmp_data)[idx] != read_bl[idx]) {
```

Comment:
Nit: technically fine for small block sizes, but the iterators will be faster since otherwise the direct indexing might involve layers of indirection iterating through `ceph::buffer::ptr` objects.